File: 91_Decode_Ways.py
class Solution:
    def isValid(self, substr):
        if len(substr) == 0 or substr[0] == "0":
            return False
        val = int(substr)
        return val > 0 and val < 27
    
    def numDecodingsUtil(self, s, left, right):
        # Base case, end depth traversal
        if right >= len(s): 
            return 0
        
        # Avoid redudant checks        
        if (left, right) in self.cache:
            return self.cache[(left, right)]
        
        if not self.isValid(s[left:right+1]):
            return 0
        else:
            # End of string reached by decision tree
            if right == len(s) - 1:
                return 1        
  
        # Recurse next decision, sum single number chosen and 2 numbers chosen pathways
        num_decodings = self.numDecodingsUtil(s, right+1, right+1) + self.numDecodingsUtil(s, right+1, right+2)
        self.cache[(left, right)] = num_decodings
        
        return num_decodings
        
    
    def numDecodings(self, s: str) -> int:        
        # Map index tuple (start, end) indices -> to its # possible decodings
        self.cache = {}
        a = self.numDecodingsUtil(s, 0, 0)
        b = self.numDecodingsUtil(s, 0, 1)
        
        return a+b

File: 91_Decode_Ways.java
class Solution {
    HashMap<Integer, Integer> cache = new HashMap<>();
    public int numDecodings(String s) {
        return numDecodingsUtil(0, s);
    }
    
    public boolean isValid(int i, int j, String s){
        if(i >= s.length()){
            return false;
        }
        String sub = s.substring(i, j);
        if(sub.charAt(0) == '0'){
            return false;
        }
        int val = Integer.valueOf(sub);
        return val > 0 && val < 27;
    }
    
    private int numDecodingsUtil(int index, String s) {
        // Lookup previous result
        if(this.cache.containsKey(index)){
            return this.cache.get(index);
        }
        // End of string reached
        if(index == s.length()){
            return 1;
        }
        // Base case, decoding is valid if this last character is valid
        if(index == s.length() - 1){
            return this.isValid(index, index+1, s) ? 1 : 0;
        }
        
        int decodings = 0;
        if(this.isValid(index, index+1, s)){
            decodings += this.numDecodingsUtil(index+1, s);
        }
        if(this.isValid(index, index+2, s)){
            decodings += this.numDecodingsUtil(index+2, s);
        }
        this.cache.put(index, decodings);
        
        return decodings;
    }
        
}